src/sinks/prometheus/exporter.rs

```diff
@@ -276,39 +278,66 @@ impl PrometheusExporter {
 impl StreamSink for PrometheusExporter {
     async fn run(&mut self, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
         self.start_server_if_needed();
-        while let Some(event) = input.next().await {
-            let item = event.into_metric();
-            let mut metrics = self.metrics.write().unwrap();
-
-            match item.kind {
-                MetricKind::Incremental => {
-                    let new = MetricEntry(item.to_absolute());
-                    if let Some(MetricEntry(mut existing)) = metrics.take(&new) {
-                        if item.value.is_set() {
-                            // sets need to be expired from time to time
-                            // because otherwise they could grow infinitelly
-                            let now = Utc::now().timestamp();
-                            let interval = now - *self.last_flush_timestamp.read().unwrap();
-                            if interval > self.config.flush_period_secs as i64 {
-                                *self.last_flush_timestamp.write().unwrap() = now;
-                                existing.reset();
-                            }
+
+        future::poll_fn(|cx| -> Poll<Result<(), ()>> {
+            let mut received = 0;
+            let mut metrics = None;
+            let mut reset_existing = None;
+
+            let result = loop {
+                match Pin::new(&mut input).poll_next(cx) {
+                    Poll::Ready(Some(event)) => {
+                        received += 1;
+
+                        if metrics.is_none() {
+                            metrics.replace(self.metrics.write().unwrap());
```

Comment:
I also suggest replacing the `unwrap()`s with `expect(…)` which both gives better error messages and internally notes why it should never fail.